mhs-reference-implementation/mhs/routing/sds.py

```diff
@@ -0,0 +1,117 @@
+import asyncio
+import logging
+import ssl
+import pathlib
+from definitions import ROOT_DIR
+import ldap3
+import mhs.routing.routing_exception as re
+
+NHS_SERVICES_BASE = "ou=services, o=nhs"
+
+MHS_OBJECT_CLASS = "nhsMhs"
+AS_OBJECT_CLASS = "nhsAs"
+MHS_PARTY_KEY = 'nhsMHSPartyKey'
+
+# These need replacing as the appropriate TLS certs are received for PTL
+PRIVATE_KEY = pathlib.Path(ROOT_DIR) / 'data' / 'certs' / 'client.key'
+LOCAL_CERT_FILE = pathlib.Path(ROOT_DIR) / 'data' / 'certs' / 'client.cert'
+CA_CERTS_FILE = pathlib.Path(ROOT_DIR) / 'data' / 'certs' / 'client.cert'
+
+mhs_attributes = [
+    'nhsEPInteractionType', 'nhsIDCode', 'nhsMhsCPAId', 'nhsMHSEndPoint', 'nhsMhsFQDN',
+    'nhsMHsIN', 'nhsMHSIsAuthenticated', 'nhsMHSPartyKey', 'nhsMHsSN', 'nhsMhsSvcIA', 'nhsProductKey',
+    'uniqueIdentifier', 'nhsMHSAckRequested', 'nhsMHSActor', 'nhsMHSDuplicateElimination',
+    'nhsMHSPersistDuration', 'nhsMHSRetries', 'nhsMHSRetryInterval', 'nhsMHSSyncReplyMode'
+]
+
+
+class SDSClient:
+
+    def __init__(self, ldap_address: str = None, tls: bool = False):
+        """
+        :param ldap_address: The IP address of the directory containing data of the the MHSs
```

Comment:
Can you add something to describe the behaviour when no `ldap_address` is provided? I.e. the client is expected to set the `connection` property themselves?

Reply:
Since the sds client now takes a connection object this no longer applies I believe.